executor/runtime/docker/docker_linux.go

```diff
@@ -116,37 +147,85 @@ func setupSystemPods(parentCtx context.Context, c *runtimeTypes.Container, cfg c
 	defer conn.Close()
 
 	// TODO: Make concurrent
-	if err := startSystemdUnit(ctx, conn, false, c.TaskID, atlasSystemdUnit); err != nil {
-		logrus.WithError(err).Error("Error starting atlas systemd unit")
+	if err := startSystemdUnit(ctx, conn, c.TaskID, c.ID, atlasServiceOpts); err != nil {
+		logrus.WithError(err).Error("Error starting atlas service")
 		return err
 	}
 	if cfg.ContainerSSHD {
-		if err := startSystemdUnit(ctx, conn, true, c.TaskID, sshdSystemdUnit); err != nil {
-			logrus.WithError(err).Error("Error starting ssh systemd unit")
+		if err := startSystemdUnit(ctx, conn, c.TaskID, c.ID, sshServiceOpts); err != nil {
+			logrus.WithError(err).Error("Error starting ssh service")
 			return err
 		}
 	}
-	if err := startSystemdUnit(ctx, conn, true, c.TaskID, metadataServiceSystemdUnit); err != nil {
-		logrus.WithError(err).Error("Error starting metadata service systemd unit")
+	if err := startSystemdUnit(ctx, conn, c.TaskID, c.ID, metadataServiceOpts); err != nil {
+		logrus.WithError(err).Error("Error starting metadata service")
 		return err
 	}
 
 	if shouldStartMetatronSync(&cfg, c) {
 		// The metatron sync service queries the metadata server, so it needs to be started after
-		if err := startSystemdUnit(ctx, conn, true, c.TaskID, metatronServiceSystemdUnit); err != nil {
-			logrus.WithError(err).Error("Error starting metatron systemd unit")
+		if err := startSystemdUnit(ctx, conn, c.TaskID, c.ID, metatronServiceOpts); err != nil {
+			logrus.WithError(err).Error("Error starting metatron service")
 			return err
 		}
 	}
 
 	return nil
 }
 
-func startSystemdUnit(ctx context.Context, conn *dbus.Conn, required bool, taskID, unitName string) error {
-	qualifiedUnitName := fmt.Sprintf("%s@%s.service", unitName, taskID)
-	l := logrus.WithField("taskID", taskID).WithField("unit", unitName)
-	l.Infof("Starting systemd unit %s", qualifiedUnitName)
+func runServiceInitCommand(ctx context.Context, log *logrus.Entry, cID string, opts serviceOpts) error {
+	if opts.initCommand == "" {
+		return nil
+	}
+
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	l := log.WithField("initCommand", opts.initCommand)
+	cmdArgs := strings.Split(fmt.Sprintf(opts.initCommandArgs, cID), " ")
 
+	l.WithField("args", cmdArgs).Infof("Running init command for %s service", opts.unitName)
+	cmd := exec.CommandContext(ctx, opts.initCommand, cmdArgs...)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	err := cmd.Run()
+	if err != nil {
+		outputStr := stdout.String()
+		l.WithField("stdout", outputStr).WithField("stderr", stderr.String()).WithError(err).Errorf("error running init command for %s service: exit code %d", opts.unitName, cmd.ProcessState.ExitCode())
```

Comment:
exit code and service should be structured fields. I would actually almost say stderr / stdout should be in the message because we can't index / query on them

Reply:
OK, I think I'll just log stdout and stderr as messages separately.

My reason for including the service in the message is that it makes it much easier to figure out what's going on when you're using `journalctl` in normal, non-json output mode.  If there's a better way of getting that info without looking through piles of JSON output, I'd love to know it!